src/Spago/PackageSet.hs

```diff
@@ -40,6 +48,112 @@ makePackageSetFile force comments = do
     else logWarn $ display $ Messages.foundExistingProject packagesPath
   Dhall.format packagesPath
 
+-- | Use the specified version of the package set (if specified).
+--   Otherwise, get the latest version of the package set if possible
+updatePackageSetVersion
+  :: forall env
+  .  (HasLogFunc env, HasGlobalCache env)
+  => Maybe Text
+  -> RIO env ()
+updatePackageSetVersion = maybe useLatestRelease useSpecificRelease
+
+useSpecificRelease
+  :: forall env
+  .  (HasLogFunc env, HasGlobalCache env)
+  => Text
+  -> RIO env ()
+useSpecificRelease tag = do
+  logDebug "Running `spago upgrade-set --tag`"
+
+  rawPackageSet <- liftIO $ Dhall.readRawExpr packagesPath
+  (org, repo, currentTag) <- case rawPackageSet of
+        Nothing -> die [ display Messages.cannotFindPackages ]
+        Just (_, expr)
+          | (current:_) <- foldMap getCurrentTag expr
+          -> pure current
+        Just _ -> die [ display Messages.cannotFindPackageImport ]
+
+  updateTag org repo currentTag tag
+  where
+    updateTag :: HasLogFunc env => Text -> Text -> Text -> Text -> RIO env ()
+    updateTag org repo currentTag specificTag =  do
+      let quotedTag = surroundQuote specificTag
+          orgRepo = org <> "/" <> repo
+      logDebug $ "Attempting to use tag for " <> display (surroundQuote orgRepo) <> ": " <> display quotedTag
+      rawPackageSet <- liftIO $ Dhall.readRawExpr packagesPath
+      case rawPackageSet of
+        Nothing -> die [ display Messages.cannotFindPackages ]
+        -- Skip the check if the tag is already being used
+        Just _
+          | currentTag == specificTag
+            -> logDebug $ "Skipping package set version upgrade, already using specified version: " <> display quotedTag
+        Just (header, expr) -> do
+          logDebug $ "Specified tag is different than current tag"
+          let newExpr = fmap (upgradeImports org repo specificTag) expr
+          logInfo $ display $ Messages.changingToSpecificPackageSet specificTag
+          resolvedExpr <- verifyPackageSetExists newExpr
+          liftIO $ Dhall.writeRawExprPostVerify packagesPath (header, newExpr) resolvedExpr
+          -- If everything is fine, refreeze the imports
+          freeze packagesPath
```

Comment:
I'd try to catch that exception simply when calling `writeExpr` here, e.g. something like this:
```suggestion
          try (liftIO $ Dhall.writeRawExpr packagesPath (header, newExpr) newExpr) >>= \case
            Right _ -> do
              -- If everything is fine, refreeze the imports
              freeze packagesPath
            Left (err :: PrettyHttpException) -> do
              -- ...rest of the code here
```

Makes sense?

Reply:
Yeah, that reads a lot better than what I had.